### ISSUE TITLE
High CPU infinite loop caused by plugin.Run() enumerating dictionary

### ISSUE THREAD

**Issue by chrisproud** [exceptionless/Exceptionless.Net#62]
What looks like a threading issue in exceptionless has just taken down one of my servers. This occurred following IIS recycling the process, so basically at app start..  I have managed to get a memory dump of the process and this was reported from the analysis:

> The following threads in w3wp.DMP are enumerating a System.Collections.Generic.Dictionary object
( 85 -snip 47 thread IDs here- 157 )
30.25% of threads blocked (49 threads)

> Multiple threads enumerating through a collection is intrinsically not a thread-safe procedure. If the dictionary object accessed by these threads is declared as static then the threads can go in an infinite loop while trying to enumerate the dictionary if one of the threads writes to the dictionary while the other threads are reading\enumerating through the same dictionary. You may also experience High CPU during this stage. For more details refer to High CPU in .NET app using a static Generic.Dictionary
http://blogs.msdn.com/b/tess/archive/2009/12/21/high-cpu-in-net-app-using-a-static-generic-dictionary.aspx

The top of the stack for each of these threads is:
>System.Collections.Generic.Dictionary`2[[System.__Canon, mscorlib],[System.__Canon, mscorlib]].FindEntry(System.__Canon)+c8 
System.Collections.Generic.Dictionary`2[[System.__Canon, mscorlib],[System.__Canon, mscorlib]].ContainsKey(System.__Canon)+9 
Exceptionless.ExceptionlessConfigurationExtensions.GetInstallId(Exceptionless.ExceptionlessConfiguration)+86 
Exceptionless.Plugins.Default.EnvironmentInfoPlugin.Run(Exceptionless.Plugins.EventPluginContext)+cd 
Exceptionless.Plugins.EventPluginManager.Run(Exceptionless.Plugins.EventPluginContext)+14b 
Exceptionless.ExceptionlessClient.SubmitEvent(Exceptionless.Models.Event, Exceptionless.Plugins.ContextData)+238 
System.Web.HttpApplication.RaiseOnError()+87 
System.Web.HttpApplication.RecordError(System.Exception)+381 
System.Web.HttpApplication.RecordError(System.Exception)+2c9 

Since Dictionary<> is thread safe on reads, i'm guessing there is a problem around reporting exceptions while exceptionless is still in startup?  (Assuming persistedClientData is only configured once in startup)

Looking in the ObservableDictionary class, my first suggestion was going to be switch Dictionary to a ConcurrenctDictionary but this class isn't supported in a portable library. I then thought maybe I could scatter locks all over the place but that won't work.

Maybe we can lock exceptionless reporting when it is in startup?

I'm happy to help fix this issue but being unfamiliar with this code i'm not sure where to go next. Any ideas how to write a test to replicate the issue?

(ps, I was going to post this on user voice as per guidelines (as this may need more discussion) but it looks inactive, maybe the guidelines need updating.)

**Comment by niemyjski** [comment]
I just updated the guidelines and seen this issue. This is not good at all. I've never seen this behavior and we log stuff on startup with our app. I'll take a look at the code in question right now.. We lock the config the second process queue or event submission occurs.

If you could send me a small sample that can reproduce that would be wonderful as I could get this under the debugger.

**Comment by niemyjski** [comment]
Another thing to do would be to enable tracing/logging https://github.com/exceptionless/Exceptionless.Net/wiki/Troubleshooting which will write out what the client is doing + log any exceptions that are submitted so you can see what error is occurring

**Comment by niemyjski** [comment]
Please feel free to join our chat room as well: https://gitter.im/exceptionless/Discuss

**Comment by niemyjski** [comment]
Maybe the test would be to run the plugin manager on multiple threads concurrently in a test.

**Comment by niemyjski** [comment]
If you can join our chat room that would be a huge help. I can also jump on with you and try and get this resolved.

**Comment by chrisproud** [comment]
Can't repo the issue. Parking this issue for now.

**Comment by coxp** [comment]
We've also hit this issue after upgrading to Exceptionless 4.0. After bringing the server down multiple times we've had to remove Exceptionless completely from our web.config (just setting enable to false didn't fix the issue) and this has fixed the issue. The infinite loop issue in our case seems to be proceeded by this error in the event viewer.

```
Exception information: 
    Exception type: KeyNotFoundException 
    Exception message: The given key was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Exceptionless.ExceptionlessConfiguration.RemovePlugin(String key)
   at Exceptionless.ExceptionlessConfiguration.AddPlugin(String key, Type pluginType)
   at Exceptionless.Mvc.ExceptionlessModule.Init(HttpApplication app)
   at System.Web.HttpApplication.InitModulesCommon()
   at System.Web.HttpApplication.InitInternal(HttpContext context, HttpApplicationState state, MethodInfo[] handlers)
   at System.Web.HttpApplicationFactory.GetNormalApplicationInstance(HttpContext context)
   at System.Web.HttpRuntime.ProcessRequestNotificationPrivate(IIS7WorkerRequest wr, HttpContext context)
```

**Comment by niemyjski** [comment]
@coxp are you still able to reproduce the issue? If so would it be possible to create a sample or could I jump on a remote meeting with you? If you could join our slack (http://slack.exceptionless.com) that would be a huge help.

Here is the thing, it looks like after we call startup (https://github.com/exceptionless/Exceptionless.Net/blob/master/src/Platforms/Exceptionless.Mvc/ExceptionlessModule.cs#L15) we could bail out if enabled is set to false, and do a lot less work, but what if you want to set everything up and then enable it in code or someplace later in the app?

@ejsmith ^^

**Comment by coxp** [comment]
No sorry, can't reproduce the issue as we're not sure what triggers it (multithreading issues are notoriously difficult to reproduce). We've removed Exceptionless for the time being as the issue causes the server to use 100% CPU and become unresponsive.

The fact it seems to fail on this line (https://github.com/exceptionless/Exceptionless.Net/blob/master/src/Exceptionless/Configuration/ExceptionlessConfiguration.cs#L379), indicates a race condition and that multiple threads are accessing the dictionary concurrently. This answer on StackOverflow to a similar issue explains how this can cause an infinite loop when using a dictionary (http://stackoverflow.com/a/33153868). 

I've never dug into the ASP.NET module system to understand where you need to be thread-safe but this is the only possible explanation for that line to not have the key in the dictionary at that point. This thread seems to imply that you will only see the issue on a production web server and not during development and that you do need to have thread-safe access to shared resources like the static ExceptionlessClient instance (http://stackoverflow.com/a/3378584 & http://stackoverflow.com/a/2619162)

**Comment by niemyjski** [comment]
A fix for this issue has been pushed in 4.0.1891. Please update to this build to resolve this issue. Thanks for bringing this to our attention.

**Comment by niemyjski** [comment]
If you have any issues or are still running into issues, please let us know.
